### PR TITLE
Fix Child Genesis

### PR DESCRIPTION
**Pull request by AdityaSripal**
- Include initial validator set in child genesis and verify it against provider client
- Return it in InitGenesis to provide Tendermint with initial validator set over ABCI

### CODE REVIEW

File: x/ccv/child/keeper/genesis.go [hunk 1]
```diff
@@ -46,6 +76,8 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state *types.GenesisState) {
 			k.SetUnbondingPacket(ctx, us.Sequence, us.UnbondingPacket)
 		}
 	}
+
+	return state.InitialValSet
```
Comment (jtremback) [inline x/ccv/child/keeper/genesis.go:80]: Am I correct in assuming that this is exactly the same as was included in the genesis.json, and is already in the format that chain startup needs?

File: x/ccv/child/keeper/genesis.go [hunk 2]
```diff
@@ -38,6 +43,31 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state *types.GenesisState) {
 		// set parent client id.
 		k.SetParentClient(ctx, clientID)
 	} else {
+		// verify that latest consensus state on parent client matches the initial validator set of restarted chain
+		// thus, IBC genesis MUST run before CCV child genesis
```
Comment (jtremback) [inline x/ccv/child/keeper/genesis.go:47]: Just wondering, could this be verified by looking at the genesis.json alone? That would eliminate that condition, not that it matters.

File: x/ccv/child/keeper/genesis.go [hunk 3]
```diff
@@ -57,7 +89,12 @@ func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 	}
 
 	if channelID, ok := k.GetParentChannel(ctx); ok {
-		gs := types.NewRestartGenesisState(channelID, nil)
+		clientID, ok := k.GetParentClient(ctx)
+		if !ok {
+			panic("parent client does not exist")
+		}
+		// ValUpdates must be filled in off-line
```
Comment (jtremback) [inline x/ccv/child/keeper/genesis.go:96]: I'm not following this comment

Comment (jtremback) [reply x/ccv/child/keeper/genesis.go:96]: Oh, they need to paste the latest provider valset updates in themselves, right? Maybe some time we can go over the exact procedure involved in doing a consumer chain genesis export and restart

**Review by jtremback — APPROVED** [review]
LGTM

File: proto/interchain_security/ccv/child/v1/genesis.proto [hunk 4]
```diff
@@ -8,23 +8,28 @@ import "ibc/lightclients/tendermint/v1/tendermint.proto";
...
 import "gogoproto/gogo.proto";
 import "interchain_security/ccv/child/v1/child.proto";
+import "tendermint/abci/types.proto";
 
 // GenesisState defines the CCV child chain genesis state
 message GenesisState {
     Params   params          = 1 [(gogoproto.nullable) = false];
-    string parent_channel_id = 2
+    string parent_client_id = 2
+    [(gogoproto.moretags) = "yaml:\"parent_client_id\""]; // empty for a completely new chain
```
Comment (jtremback) [inline proto/interchain_security/ccv/child/v1/genesis.proto:17]: Whitespace?

File: x/ccv/child/keeper/genesis.go (hunk 1 above)
Comment (AdityaSripal) [reply x/ccv/child/keeper/genesis.go:80]: Correct. This is assuming the genesis file just includes the `[]abci.ValidatorUpdate` which is exactly what is expected by Tendermint

File: x/ccv/child/keeper/genesis.go (hunk 2 above)
Comment (AdityaSripal) [reply x/ccv/child/keeper/genesis.go:47]: You can do it for the initial child chain genesis on `ValidateGenesis` which I do.

For restart child chain genesis, we need to read the state of the IBC module. There isn't a way for ValidateGenesis to look at another module's genesis. So we need to look at the already initialized IBC state in `InitGenesis` to do our verification.

File: x/ccv/child/keeper/genesis.go [hunk 5]
```diff
@@ -38,6 +43,31 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state *types.GenesisState) {
...
+
+		// ensure that initial validator set is same as initial consensus state on provider client.
+		// this will be verified by provider module on channel handshake.
+		vals, err := tmtypes.PB2TM.ValidatorUpdates(state.InitialValSet)
+		if err != nil {
+			panic(err)
+		}
+		valSet := tmtypes.NewValidatorSet(vals)
+
+		if !bytes.Equal(tmConsState.NextValidatorsHash, valSet.Hash()) {
```
Comment (mpoke) [inline x/ccv/child/keeper/genesis.go:65]: This check requires that the initial validator set in the GenesisState matches the one from the latest consensus state stored in the provider client. This consensus state depends on the last ClientUpdate before the GenesisState was exported. How can an off-chain process identify the validator set of that consensus state?

**Pull request merged**